plugins/uefi-dbx/fu-dbxtool.c

```diff
@@ -0,0 +1,187 @@
+	if (!g_option_context_parse (context, &argc, &argv, &error)) {
+		/* TRANSLATORS: the user didn't read the man page */
+		g_print ("%s: %s\n", _("Failed to parse arguments"),
+			 error->message);
+		return EXIT_FAILURE;
+	}
+
+	/* set verbose? */
+	if (verbose) {
+		g_setenv ("G_MESSAGES_DEBUG", "all", FALSE);
+	} else {
+		g_log_set_handler (G_LOG_DOMAIN, G_LOG_LEVEL_DEBUG,
+				   fu_util_ignore_cb, NULL);
+	}
+
+	/* nothing specified */
+	if (!action_apply && !action_list) {
+		g_autofree gchar *tmp = NULL;
+		tmp = g_option_context_get_help (context, TRUE, NULL);
+		/* TRANSLATORS: user did not tell the tool what to do */
+		g_printerr ("%s\n\n%s", _("No action specified!"), tmp);
+		return EXIT_FAILURE;
+	}
+
+	/* load dbx on the system */
+	if (!fu_efivar_get_data (FU_EFIVAR_GUID_SECURITY_DATABASE, "dbx",
+				 &buf_system, &bufsz, NULL, &error)) {
+		/* TRANSLATORS: could not read existing data */
+		g_printerr ("%s: %s\n", _("Failed to load EFI dbx"), error->message);
+		return EXIT_FAILURE;
+	}
+	/* TRANSLATORS: reading existing dbx from the system */
+	g_print ("%s\n", _("Parsing current dbx…"));
+	dbx_system = fu_uefi_dbx_file_new (buf_system, bufsz,
+					   FU_UEFI_DBX_FILE_PARSE_FLAGS_NONE,
+					   &error);
+	if (dbx_system == NULL) {
+		/* TRANSLATORS: could not read existing data format */
+		g_printerr ("%s: %s\n", _("Failed to parse EFI dbx"), error->message);
+		return EXIT_FAILURE;
+	}
+	checksums_system = fu_uefi_dbx_file_get_checksums (dbx_system);
+
+	/* show existing entries */
+	if (action_list) {
+		for (guint i = 0; i < checksums_system->len; i++) {
+			const gchar *checksum = g_ptr_array_index (checksums_system, i);
+			g_print ("%03u: %s\n", i + 1, checksum);
+		}
+		return EXIT_SUCCESS;
+	}
+
+	/* get update dbx */
+	if (dbxfile == NULL) {
+		/* TRANSLATORS: user did not include a filename parameter */
+		g_printerr ("%s\n", _("Filename required"));
+		return EXIT_FAILURE;
+	}
+	if (!g_file_get_contents (dbxfile, (gchar **) &buf_update, &bufsz, &error)) {
+		/* TRANSLATORS: filename could not be read */
+		g_printerr ("%s %s: %s", _("Failed to load file"), dbxfile, error->message);
+		return EXIT_FAILURE;
+	}
+	/* TRANSLATORS: reading new dbx from the update */
+	g_print ("%s\n", _("Parsing dbx update…"));
+	dbx_update = fu_uefi_dbx_file_new (buf_update, bufsz,
+					   FU_UEFI_DBX_FILE_PARSE_FLAGS_IGNORE_HEADER,
+					   &error);
+	if (dbx_update == NULL) {
+		/* TRANSLATORS: dbx could not be read */
+		g_printerr ("%s %s: %s", _("Failed to parse file"), dbxfile, error->message);
+		return EXIT_FAILURE;
+	}
+
+	/* check this is a newer dbx update */
+	checksums_update = fu_uefi_dbx_file_get_checksums (dbx_update);
+	if (!force && fu_uefi_dbx_file_has_checksums (dbx_system, checksums_update)) {
+		/* TRANSLATORS: same or newer update already applied */
+		g_printerr ("%s\n", _("Cannot apply update as this dbx update has already been applied."));
+		return EXIT_FAILURE;
+	}
+
```

Comment:
I guess assuming this is merged first, the checks added as part of https://github.com/fwupd/fwupd/pull/2317 should slip right in here.

Reply:
Exactly. I think #2317 needs a lot more work, and I think a standalone dbxtool is useful without the extra tests.